### PR TITLE
feat: Create private chat room per event

### PR DESCRIPTION
**Pull request by MeghalBisht**
1. Separated Login and Registration functions for Rocket chat.
2. Private room are now created per event instead of system wide single room
3. Bot user with email openeventbot@openevent.com is owner of all rooms and would add users to a room. (as  users cannot add themselves inside private rooms)

#### Checklist

- [ ] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia) and my PR follows them.
- [ ] My branch is up-to-date with the Upstream `development` branch.
- [ ] The unit tests pass locally with my changes 
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

- [ ] All the functions created/modified in this PR contain relevant docstrings.

### CODE REVIEW

**Comment by MeghalBisht** [comment]
Screenshot from 2021-03-31 04-40-56

File: app/api/schema/settings.py [hunk 1]
```diff
@@ -270,6 +270,9 @@ class Meta:
     smtp_encryption = fields.Str(allow_none=True)  # Can be tls, ssl, none
 
     rocket_chat_registration_secret = fields.Str(allow_none=True)
+    rocket_bot_email = fields.Str(allow_none=True)
+    rocket_bot_password = fields.Str(allow_none=True)
+    rocket_bot_name = fields.Str(allow_none=True)
```
Comment (iamareebjamal) [inline app/api/schema/settings.py:275]: Why in schema? Shouldn't these be privaate internal fields?

**Comment by MeghalBisht** [comment]
ready for review! @iamareebjamal

File: app/api/schema/events.py [hunk 2]
```diff
@@ -83,6 +83,7 @@ def validate_timezone(self, data, original_data):
     is_billing_info_mandatory = fields.Bool(default=False)
     is_donation_enabled = fields.Bool(default=False)
     is_chat_enabled = fields.Bool(default=False)
+    chat_room_id = fields.Str(allow_none=True)
```
Comment (iamareebjamal) [inline app/api/schema/events.py:86]: Shouldn't be needed, this is internal and not editable or viewable

Comment (MeghalBisht) [reply app/api/schema/events.py:86]: room id is needed to add a user into a room. From where will it be accessed?

Comment (iamareebjamal) [reply app/api/schema/events.py:86]: Is it needed in frontend?

Comment (iamareebjamal) [reply app/api/schema/events.py:86]: If so, this should be dumps_only

Comment (MeghalBisht) [reply app/api/schema/events.py:86]: not in frontend, but for add user api call

**Comment by iamareebjamal** [comment]
@MeghalBisht I still can't see where in frontend do we use chat_room_id? We just use event.name + event.identifier?

Which is still not a good solution IMO

File: app/api/chat/rocket_chat.py [hunk 3]
```diff
@@ -24,76 +27,165 @@ def __init__(self, message, code=CODES.UNHANDLED, response=None) -> None:
...
+        db.session.commit()
+        add_in_room(event, user_id, api_url, login_url)
+
+
+def add_in_room(event: Event, rocket_user_id, api_url='', login_url=''):
+    bot_info = check_or_create_bot(event, api_url, login_url)
+    data = get_rocket_chat_token(bot_info, event, False, True)
+    bot_token = data['token']
+    bot_id = data['res']['data']['userId']
+    room_info = {'roomId': event.chat_room_id, 'userId': rocket_user_id}
```
Comment (MeghalBisht) [inline app/api/chat/rocket_chat.py:146]: chat room id is used here

Comment (iamareebjamal) [reply app/api/chat/rocket_chat.py]: That's not frontend, no need to add something in schema which is not used in frontend. It gets exposed in the API

Comment (MeghalBisht) [reply app/api/chat/rocket_chat.py]: Ok, but how will room id be given here then? as room id is needed to add a user into a room.

Comment (iamareebjamal) [reply app/api/chat/rocket_chat.py]: Schema and Models are different things. Schema is for API, and model is for code. No change would be needed here if chat_room_id is not in the schema. Schema exposes stuff or takes a field as an input publicly through an API

**Comment by MeghalBisht** [comment]
currently special characters are not included in name.
What can be a better way to rename a room in such a case?

File: app/models/event.py [hunk 4]
```diff
@@ -61,6 +61,7 @@ class Privacy:
     is_promoted = db.Column(db.Boolean, default=False, nullable=False)
     is_demoted = db.Column(db.Boolean, default=False, nullable=False)
     is_chat_enabled = db.Column(db.Boolean, default=False, nullable=False)
+    chat_room_name = db.Column(db.String)
```
Comment (iamareebjamal) [inline app/models/event.py:64]: No need for a field

File: app/models/event.py [hunk 5]
```diff
@@ -385,6 +386,11 @@ def calc_revenue(self, start=None, end=None):
             or 0
         )
 
+    @property
+    def rocket_room_name(self):
+        self.chat_room_name = self.name.replace(' ', '-') + '-' + self.identifier
+        return self.chat_room_name
```
Comment (iamareebjamal) [inline app/models/event.py:392]: Why assigning to a field and returning. Doesn't make sense

**Comment by iamareebjamal** [comment]
> What can be a better way to rename a room in such a case?

Use a regex

File: app/models/event.py (hunk 4 above)
Comment (MeghalBisht) [reply app/models/event.py:64]: what about frontend parameter?

File: app/models/event.py [hunk 6]
```diff
@@ -388,8 +388,7 @@ def calc_revenue(self, start=None, end=None):
 
     @property
     def rocket_room_name(self):
```
Comment (iamareebjamal) [inline app/models/event.py:390]: Rename to chat_room_name and remove the field

File: app/models/event.py (hunk 4 above)
Comment (iamareebjamal) [reply app/models/event.py:64]: That's backed by the property

File: app/api/schema/events.py [hunk 7]
```diff
@@ -83,7 +83,6 @@ def validate_timezone(self, data, original_data):
     is_billing_info_mandatory = fields.Bool(default=False)
     is_donation_enabled = fields.Bool(default=False)
     is_chat_enabled = fields.Bool(default=False)
-    chat_room_name = fields.Str(allow_none=True)
```
Comment (iamareebjamal) [inline app/api/schema/events.py]: If you remove it from schema, how will frontend access it?

Comment (iamareebjamal) [reply app/api/schema/events.py]: This should be dumps_only field

Comment (MeghalBisht) [reply app/api/schema/events.py]: yes I saw, will push the changes with regex commit

File: app/api/schema/events.py [hunk 8]
```diff
@@ -83,6 +83,7 @@ def validate_timezone(self, data, original_data):
     is_billing_info_mandatory = fields.Bool(default=False)
     is_donation_enabled = fields.Bool(default=False)
     is_chat_enabled = fields.Bool(default=False)
+    chat_room_name = fields.Str(dump_only=True, allow_none=True)
```
Comment (iamareebjamal) [inline app/api/schema/events.py:86]:
```suggestion
    chat_room_name = fields.Str(dump_only=True)
```

**Pull request merged**